plugins/modules/slack.py

```diff
@@ -431,6 +481,143 @@ def do_notify_slack(module, domain, token, payload):
+            params["snippet_type"] = file_upload.get("snippet_type")
+        params = urlencode(params)
+        response, info = fetch_url(
+            module, "%s?%s" % (url, params), headers=headers, method="GET"
+        )
+        if info["status"] != 200:
+            module.fail_json(
+                msg="Error retrieving upload URL: %s (HTTP %s)" % (info['msg'], info['status'])
+            )
+        try:
+            upload_url_data = json.load(response)
+        except ValueError:
+            module.fail_json(
+                msg="The Slack API response is not valid JSON: %s" % response.read()
+            )
+        if not upload_url_data.get("ok"):
+            module.fail_json(
+                msg="Failed to retrieve upload URL: %s" % upload_url_data.get('error')
+            )
+        upload_url = upload_url_data["upload_url"]
+        file_id = upload_url_data["file_id"]
+        # Step 2: Upload file content
+        try:
+            with open(file_path, "rb") as file:
+                file_content = file.read()
+            response, info = fetch_url(
+                module,
+                upload_url,
+                data=file_content,
+                headers={"Content-Type": "application/octet-stream"},
+                method="POST",
+            )
+            if info["status"] != 200:
+                module.fail_json(
+                    msg="Error during file upload: %s (HTTP %s)" % (info['msg'], info['status'])
+                )
+        except IOError:
+            module.fail_json(msg="The file %s is not found." % file_path)
+        # Step 3: Complete upload
+        complete_url = SLACK_COMPLETE_UPLOAD_EXTERNAL
+        files_dict = {
+            "files": [
+                {
+                    "id": file_id,
+                }
+            ],
+            "channel_id": get_channel_id(module, token, channel),
+        }
+        if file_upload.get("title"):
+            files_dict["files"][0]["title"] = file_upload.get("title")
+        if file_upload.get("initial_comment"):
+            files_dict["initial_comment"] = file_upload.get("initial_comment")
+        if file_upload.get("thread_ts"):
+            files_dict["thread_ts"] = file_upload.get("thread_ts")
+        files_data = json.dumps(files_dict)
+        headers = {
+            "Authorization": "Bearer " + token,
+            "Content-Type": "application/json",
+        }
+        try:
+            response, info = fetch_url(
+                module, complete_url, data=files_data, headers=headers, method="POST"
+            )
+            if info["status"] != 200:
+                module.fail_json(
+                    msg="Error during upload completion: %s (HTTP %s)" % (info['msg'], info['status'])
+                )
+            upload_url_data = json.load(response)
+        except ValueError:
+            module.fail_json(
+                msg="The Slack API response is not valid JSON: %s" % response.read()
```

Comment:
At this point the response has likely already been read. You should read it first and *then* use the result both to decode JSON and to report the error here, otherwise this reporting here won't work.

Reply:
```suggestion
        try:
            upload_url_data = json.load(response)
        except ValueError:
            response_content = response.read()
            module.fail_json(msg="The Slack API response is not valid JSON: %s" % response_content)
```

Thanks for the feedback! You're right.